### DISCUSSION TITLE
Canceling a specific instance of a job

### DISCUSSION THREAD

**Discussion by tarockx** [quartznet/quartznet#2028] [Q&A] [answered]
Hi everyone

I have an application where I have jobs that can be triggered on-demand and need to be cancelled on-demand.

I'm triggering the job with this simple code that works without problems:

```
var trigger = TriggerBuilder.Create()
  .WithIdentity(myTriggerKey)
  .ForJob(myJobKey)
  .StartNow()
  .Build();
myScheduler.ScheduleJob(trigger);
```

However, I am having trouble implementing the job cancellation part.
There are two methods mentioned in the API docs, but none of them seems suitable to my case:

- The `Interrupt(JobKey jobKey, CancellationToken cancellationToken = default)` method interrupts ALL jobs of a certain type, but I only want to interrupt a specific instance of the job
- The `Interrupt(string fireInstanceId, CancellationToken cancellationToken = default)` method seems to be able to interrupt a specific instance of the job, however you have to provide the `fireIntanceId` parameter, which from what I can tell is only available INSIDE the job (inside the `IJobExecutionContext` parameter that is passed to the job's `.Execute() method`). However, this seems quite backwards to me, since typically you would want to cancel a job from OUTSIDE, i.e.: where the `IScheduler` instance is available rather than from inside the job itself. In other words:
```
myScheduler.Interrupt(fireInstanceId) // <-- How do I get the fireInstanceId to pass here?
```

To me it would be much more logical to be able to either:
- specifying the `fireInstanceId` manually in the trigger, so I can store it when I trigger the job and recover it later to cancel it
- being able to cancel a job by using the `TriggerKey`, like we can do with the `JobKey`

But I don't see any way to do this in the current implementation of Quartz.

Is there a straightforward solution to this problem? At the moment I am working around this by having a static `Dictionary<TriggerKey, string>`. When I create the trigger, I make sure the TriggerKey I use is unique so I can use it later to cancel the job. When the job starts, as the first instruction in the `.Execute()` method I recover the `FireInstanceId` and put it in the dictionary, matched with the corresponding `TriggerKey`. Later, I can use the trigger key that I previously stored to recover the `FireInstanceId` and finally request the job's interruption by calling the `.Interrupt()` method of the scheduler. 

This works, but it's a very clunky/hacky solution: not only it requires more steps and moving parts than it would seem necessary, but you also need to make sure you clean up the dictionary whenever an instance of a job ends, otherwise it will fill with stale objects. Moreover, it doesn't provide a way to cancel a job that has been triggered but isn't executing yet (for example because there aren't any threads available in the thread pool to start it) because the `FireInstanceId` can only be "injected" in the dictionary when the job actually starts executing.

Is there any better ways to do this?

**Comment by lahma** [comment, marked as answer]
Interrupting with `fireInstanceId` is not cluster aware. But you can get currentl executiing jobs from scheduler using `GetCurrentlyExecutingJobs()` which returns list of `IJobExecutionContext`s which have the fire instance id you need for that overload.

**Reply by tarockx** [reply]
That's ok, I'm not running in a cluster. Thanks, I'll try your solution
